src/main/java/org/team4159/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,36 +1,186 @@
 package org.team4159.frc.robot.subsystems;
 
+import com.ctre.phoenix.motorcontrol.FeedbackDevice;
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+import com.ctre.phoenix.sensors.PigeonIMU;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+
 import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.geometry.Pose2d;
+import edu.wpi.first.wpilibj.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.geometry.Translation2d;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveWheelSpeeds;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
-import com.ctre.phoenix.motorcontrol.NeutralMode;
-import com.ctre.phoenix.motorcontrol.can.TalonSRX;
-import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
-
 import static org.team4159.frc.robot.Constants.*;
 
 public class Drivetrain extends SubsystemBase {
-  private SpeedControllerGroup left_talons;
-  private SpeedControllerGroup right_talons;
+  private TalonFX left_front_falcon, left_rear_falcon, right_front_falcon, right_rear_falcon;
+  private SpeedControllerGroup left_falcons;
+  private SpeedControllerGroup right_falcons;
+
+  private DifferentialDrive differential_drive;
+
+  private DifferentialDriveOdometry odometry;
+  private PigeonIMU pigeon;
+
+  private boolean is_oriented_forward = true;
+
+  private double dx = 0;
+  private double dy = 0;
+  private double prev_magnitude = 0;
 
   public Drivetrain() {
-    left_talons = new SpeedControllerGroup(
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.LEFT_FRONT_TALON)),
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.LEFT_REAR_TALON)));
-    right_talons = new SpeedControllerGroup(
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.RIGHT_FRONT_TALON)),
-            (WPI_TalonSRX) configureTalonSRX(new WPI_TalonSRX(CAN_IDS.RIGHT_REAR_TALON)));
+    left_front_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.LEFT_FRONT_FALCON_ID));
+    left_rear_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.LEFT_REAR_FALCON_ID));
+    right_front_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.RIGHT_FRONT_FALCON_ID));
+    right_rear_falcon = configureTalonFX(new WPI_TalonFX(CAN_IDS.RIGHT_REAR_TALON_ID));
+
+    left_front_falcon.configSelectedFeedbackSensor(FeedbackDevice.IntegratedSensor);
+    right_front_falcon.configSelectedFeedbackSensor(FeedbackDevice.IntegratedSensor);
+
+    // setSensorPhase isn't working
+
+    left_falcons = new SpeedControllerGroup(
+      (WPI_TalonFX) left_front_falcon,
+      (WPI_TalonFX) left_rear_falcon);
+    right_falcons = new SpeedControllerGroup(
+      (WPI_TalonFX) right_front_falcon,
+      (WPI_TalonFX) right_rear_falcon);
+    left_falcons.setInverted(true);
```

Comment:
Group the directions together, instead of having interleaving the left and right talons

Reply:
Not sure what you mean here